cpp/src_prims/sparse/distance/coo_spmv.cuh

```diff
@@ -0,0 +1,306 @@
+  value_idx active_chunk_size = min(chunk_size * tpb, nnz_b - ind_offset);
+
+  int tid = threadIdx.x;
+  int warp_id = tid / raft::warp_size();
+
+  // compute id relative to current warp
+  unsigned int lane_id = tid & (raft::warp_size() - 1);
+  value_idx ind = ind_offset + threadIdx.x;
+
+  if (cur_row_a > m || cur_chunk_offset > n_blocks_per_row) return;
+  if (ind >= nnz_b) return;
+
+  extern __shared__ char smem[];
+
+  value_idx *offsets_a = (value_idx *)smem;
+  kv_t *A = (kv_t *)(smem + (2 * sizeof(value_idx)));
+  typename warp_reduce::TempStorage *temp_storage =
+    (typename warp_reduce::TempStorage *)(A + buffer_size);
+
+  if (tid == 0) {
+    offsets_a[0] = indptrA[cur_row_a];
+    offsets_a[1] = indptrA[cur_row_a + 1];
+
+    // Create dense vector A and populate with 0s
+    memset(A, 0, buffer_size * sizeof(kv_t));
+  }
+
+  __syncthreads();
+
+  value_idx start_offset_a = offsets_a[0];
+  value_idx stop_offset_a = offsets_a[1];
+
+  // Convert current row vector in A to dense
+  for (int i = tid; i < (stop_offset_a - start_offset_a); i += blockDim.x) {
+    A[indicesA[start_offset_a + i]] = dataA[start_offset_a + i];
+  }
+
+  __syncthreads();
+
+  value_idx cur_row_b = -1;
+  value_t c = 0.0;
+
+  auto warp_red = warp_reduce(*(temp_storage + warp_id));
+
+  if (tid < active_chunk_size) {
+    cur_row_b = rowsB[ind];
+    value_t a_col = A[indicesB[ind]];
+    if (!rev || a_col == 0.0) c = reduce_func(a_col, dataB[ind]);
+  }
+
+  // loop through chunks in parallel, reducing when a new row is
+  // encountered by each thread
+  for (int i = tid; i < active_chunk_size; i += blockDim.x) {
+    value_idx ind_next = ind + blockDim.x;
+    value_idx next_row_b = -1;
+
+    if (i + blockDim.x < active_chunk_size) next_row_b = rowsB[ind_next];
+
+    if (next_row_b != cur_row_b) {
+      unsigned int peer_group = get_peer_group(cur_row_b);
+      bool is_leader = get_lowest_peer(peer_group) == lane_id;
+
+      value_t v = warp_red.HeadSegmentedReduce(c, is_leader, accum_func);
+
+      // thread with lowest lane id among peers writes out
+      if (is_leader && v != 0.0) {
+        value_idx idx =
+          !rev ? cur_row_a * n + cur_row_b : cur_row_b * m + cur_row_a;
+        write_func(out + idx, v);
+      }
+      c = 0.0;
+    }
+
+    if (next_row_b != -1) {
+      ind = ind_next;
+      value_t a_col = A[indicesB[ind]];
+      if (!rev || a_col == 0.0)
+        c = accum_func(c, reduce_func(a_col, dataB[ind]));
+      cur_row_b = next_row_b;
+    }
+  }
+}
+
+/**
+ * Computes the maximum number of columns that can be stored
+ * in shared memory in dense form with the given block size
+ * and precision.
+ * @return the maximum number of columns that can be stored in smem
+ */
+template <typename value_idx, typename value_t, int tpb = 1024>
+inline int max_cols_per_block() {
+  // compute max shared mem to use
+  return (raft::getSharedMemPerBlock() - (2 * sizeof(value_idx)) -
```

Comment:
Could you write a small quick example here so that the calculation is more apparent?